metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1199,6 +1213,22 @@ public void processBatchEndOffset(long offset) {
         });
     }
 
+    @Override
+    public CompletableFuture<AllocateProducerIdsResponseData> allocateProducerIds(
+            AllocateProducerIdsRequestData request) {
+        return appendWriteEvent("allocateProducerIds",
+            () -> producerIdControlManager.generateNextProducerId(request.brokerId(), request.brokerEpoch()))
+                .thenApply(resultOrError -> {
```

Comment:
I can see the reasoning behind doing it this way, but for all the other RPCs we've just been letting the future complete as an error, and making the caller handle it.  One issue with changing the pattern, I suppose, is that not all the controller functions return an RPC data structure that allows setting an error. So let's keep the current pattern here for now, where caller has to handle the future completing exceptionally.

Reply:
Sounds fine. I'll change the return type of generateNextProducerId to just be `ControllerResult<ProducerIdBlock>` and throw ApiExceptions directly instead.